pallets/process-validation/src/lib.rs

```diff
@@ -79,49 +89,120 @@ pub mod pallet {
+        // process is already disabled
+        AlreadyDisabled,
+        // process not found for this versiion
+        InvalidVersion,
     }
 
     // The pallet's dispatchable functions.
     #[pallet::call]
     impl<T: Config> Pallet<T> {
-        // TODO: implement create_process with correct parameters and impl
         #[pallet::weight(T::WeightInfo::create_process())]
-        pub(super) fn create_process(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            // Check it was signed and get the signer
+        pub(super) fn create_process(
+            origin: OriginFor<T>,
+            id: T::ProcessIdentifier,
+            restrictions: Vec<Restriction>,
+        ) -> DispatchResultWithPostInfo {
             T::CreateProcessOrigin::ensure_origin(origin)?;
+            let new_version: i32 = Pallet::<T>::_update_version(id.clone()).unwrap();
+            Pallet::<T>::_persist_process(&id, &new_version, restrictions.clone());
+
+            // wrap event data into a struct or some other data structure?
+            Self::deposit_event(Event::ProcessCreated(id, new_version, restrictions, new_version == 1));
 
-            Self::deposit_event(Event::ProcessCreated);
-            Ok(().into())
+            return Ok(().into());
         }
 
-        // TODO: implement disable_process with correct parameters and impl
         #[pallet::weight(T::WeightInfo::disable_process())]
-        pub(super) fn disable_process(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub(super) fn disable_process(
+            origin: OriginFor<T>,
+            id: T::ProcessIdentifier,
+            version: i32,
+        ) -> DispatchResultWithPostInfo {
             T::DisableProcessOrigin::ensure_origin(origin)?;
-            Self::deposit_event(Event::ProcessDisabled);
-            Ok(().into())
+
+            ensure!(
+                <ProcessModel<T>>::contains_key(&id, version),
+                Error::<T>::NonExistingProcess,
+            );
+
+            // TODO check if there is any version for this process
+            ensure!(<VersionModel<T>>::contains_key(&id), Error::<T>::InvalidVersion,);
+
+            // Question for Matt, whether status is already disable should be an error or not
+            // also for the above ensure! macros
+            Pallet::<T>::_disable_process(&id, &version)?;
+
+            Self::deposit_event(Event::ProcessDisabled(id, version));
+            return Ok(().into());
+        }
+    }
+
+    // helper methods
+    impl<T: Config> Pallet<T> {
+        pub fn _get_version(id: T::ProcessIdentifier) -> i32 {
+            return <VersionModel<T>>::get(&id);
+        }
+
+        pub fn _update_version(id: T::ProcessIdentifier) -> Result<i32, ()> {
+            let version: i32 = Pallet::<T>::_get_version(id.clone());
+            let exists: bool = <ProcessModel<T>>::contains_key(id.clone(), version);
+            let new_version: i32 = match exists {
+                true => version,
```

Comment:
sorry if I'm misunderstanding the code but I'd expect these to be the other way round, if the process already exists version number increases?

Reply:
@jonmattgray I know, this is something I would need some clarity, looks like it works other way around not sure why tho. Need to do some more invstigation tho

Reply:
it was a bug, updated